File: design_patterns/C++/Behavioral/visitor.cxx
#include <iostream>

class Monkey;
class Lion;
class Dolphin;

// Visitor
class AnimalOperation {
  public:
	virtual void visitMonkey(Monkey* monkey) = 0;
	virtual void visitLion(Lion* lion) = 0;
	virtual void visitDolphin(Dolphin* dolphin) = 0;
};

// Visitee
class Animal {
  public:
	virtual void accept(AnimalOperation* operation) = 0;
};

class Monkey: public Animal {
  public:
	void shout() {
		std::cout << "Ooh oo aa aa!" << std::endl;
	}

	void accept(AnimalOperation* operation) {
		operation->visitMonkey(this);
	}
};

class Lion: public Animal {
  public:
	void roar() {
		std::cout << "Roaaar!" << std::endl;
	}

	void accept(AnimalOperation* operation) {
		operation->visitLion(this);
	}
};

class Dolphin: public Animal {
  public:
	void speak() {
		std::cout << "Tuut tuttu tuutt!" << std::endl;
	}

	void accept(AnimalOperation* operation) {
		operation->visitDolphin(this);
	}
};

class Speak: public AnimalOperation {
  public:
	void visitMonkey(Monkey* monkey) {
		monkey->shout();
	}

	void visitLion(Lion* lion) {
		lion->roar();
	}

	void visitDolphin(Dolphin* dolphin) {
		dolphin->speak();
	}
};

class Jump: public AnimalOperation {
  public:
	void visitMonkey(Monkey* monkey) {
		std::cout << "Jumped 20 feet high! on to the tree!" << std::endl;
	}

	void visitLion(Lion* lion) {
		std::cout << "Jumped 7 feet! Back on the ground!" << std::endl;
	}

	void visitDolphin(Dolphin* dolphin) {
		std::cout << "Walked on water a little and disappeared" << std::endl;
	}
};

int main() {
	auto monkey = new Monkey();
	auto lion = new Lion();
	auto dolphin = new Dolphin();
	auto speak = new Speak();
	auto jump = new Jump();
	monkey->accept(speak);
	lion->accept(speak);
	dolphin->accept(speak);
	// =================
	// Ooh oo aa aa!
	// Roaaar!
	// Tuut tuttu tuutt!
	// =================
	monkey->accept(jump);
	lion->accept(jump);
	dolphin->accept(jump);
	// ========================================
	// Jumped 20 feet high! on to the tree!
	// Jumped 7 feet! Back on the ground!
	// Walked on water a little and disappeared
	// ========================================
	return 0;
}

File: design_patterns/Python/Behavioral/visitor.py
class Animal:
    # visitee

    def accept(operation):
        raise NotImplementedError


class AnimalOperation:
    # visitor

    def visit_monkey(monkey):
        raise NotImplementedError

    def visit_lion(lion):
        raise NotImplementedError

    def visit_dolphin(dolphin):
        raise NotImplementedError


class Monkey(Animal):

    def shout(self):
        print('Ooh oo aa aa!')

    def accept(self, operation):
        operation.visit_monkey(self)


class Lion(Animal):

    def roar(self):
        print('Roaaar!')

    def accept(self, operation):
        operation.visit_lion(self)


class Dolphin(Animal):

    def speak(self):
        print('Tuut tuttu tuutt!')

    def accept(self, operation):
        operation.visit_dolphin(self)


class Speak(AnimalOperation):

    def visit_monkey(self, monkey):
        monkey.shout()

    def visit_lion(self, lion):
        lion.roar()

    def visit_dolphin(self, dolphin):
        dolphin.speak()


class Jump(AnimalOperation):

    def visit_monkey(self, monkey):
        print('Jumped 20 feet high! on to the tree!')

    def visit_lion(self, lion):
        print('Jumped 7 feet! Back on the ground!')

    def visit_dolphin(self, dolphin):
        print('Walked on water a little and disappeared.')


if __name__ == '__main__':

    monkey = Monkey()
    lion = Lion()
    dolphin = Dolphin()

    speak = Speak()
    jump = Jump()

    monkey.accept(speak)  # Ooh oo aa aa!
    monkey.accept(jump)  # Jumped 20 feet high! on to the tree!

    lion.accept(speak)  # Roaaar!
    lion.accept(jump)  # Jumped 7 feet! Back on the ground!

    dolphin.accept(speak)  # Tuut tuttu tuutt!
    dolphin.accept(jump)  # Walked on water a little and disappeared.
